d2l/torch.py

```diff
@@ -3071,6 +3071,136 @@ def predict_seq2seq(net, src_sentence, src_vocab, tgt_vocab, num_steps,
+            Above text/description is copied from https://gym.openai.com/envs/FrozenLake-v0/
+            and how to process env.P.items is adpated from https://sites.google.com/view/deep-rl-bootcamp/labs
+
+    '''
+    import gym
+    env = gym.make('FrozenLake-v1', is_slippery=False)
+    env.seed(seed)
+    env.action_space.np_random.seed(seed)
+    env.action_space.seed(seed)
+    env_info = {}
+    env_info['desc'] = env.desc  # 2D array specifying what each grid item means
+    env_info['num_states']     = env.nS # number of observations/states or obs/state dim
+    env_info['num_actions']    = env.nA # number of actions or action dim
+    # define indices for (transition probability, nextstate, reward, done) tuple
+    env_info['trans_prob_idx'] = 0 # index of transition probability entry
+    env_info['nextstate_idx']  = 1 # index of next state entry
+    env_info['reward_idx']     = 2 # index of reward entry
+    env_info['done_idx']       = 3 # index of done entry
+    env_info['mdp'] = {}
+    env_info['env'] = env
+
+    for (s, others) in env.P.items():
+        # others(s) = {a0: [ (p(s'|s,a0), s', reward, done),...], a1:[...], ...}
+
+        for (a, pxrds) in others.items():
+            # pxrds is [(p1,next1,r1,d1),(p2,next2,r2,d2),..].
+            # e.g. [(0.3, 0, 0, False), (0.3, 0, 0, False), (0.3, 4, 1, False)]
+            env_info['mdp'][(s,a)] = pxrds
+
+    return env_info
+
+def make_env(name ='', seed=0):
+    """Defined in :numref:`sec_code_rl`"""
+    # Input parameters:
+    # name: specifies a gym environment.
+    # For Value iteration, only FrozenLake-v1 is supported.
+    if name == 'FrozenLake-v1':
+        return FrozenLake(seed)
+
+    else:
+        raise ValueError("%s env is not supported in this Notebook")
+
+def show_value_function_progress(env_desc, V, pi):
+    """Defined in :numref:`sec_code_rl`"""
+    # This function visualizes how value and policy changes over time.
+    # V:  [num_iters, num_states]
+    # pi: [num_iters, num_states]
+    # How to visualize value function is adapted (but changed) from: https://sites.google.com/view/deep-rl-bootcamp/labs
+
+    num_iters = V.shape[0]
+    fig, ax = plt.subplots(figsize=(15, 15))
+
+    for k in range(V.shape[0]):
+        plt.subplot(4, 4, k + 1)
+        plt.imshow(V[k].reshape(4,4), cmap="bone")
+        ax = plt.gca()
+        ax.set_xticks(np.arange(0, 5)-.5, minor=True)
+        ax.set_yticks(np.arange(0, 5)-.5, minor=True)
+        ax.grid(which="minor", color="w", linestyle='-', linewidth=3)
+        ax.tick_params(which="minor", bottom=False, left=False)
+        ax.set_xticks([])
+        ax.set_yticks([])
+
+        # LEFT action: 0.
+        # DOWN action: 1.
+        # RIGHT action: 2.
+        # UP action: 3.
+        action2dxdy = {0: (-.25, 0),
+                       1: (0, .25),
+                       2: (0.25,0),
+                       3: (-.25, 0)
+                      }
```

Comment:
Make these lines more compact. Otherwise we end up with very big code blocks, which are not printing-friendly for books

Reply:
updated. let me know whether or not it looks better!